### DISCUSSION TITLE
Improving the `exec/start` and `container/exec` stream protocols

### DISCUSSION THREAD

**Discussion by corhere** [moby/moby#45414] [Ideas]
The protocol for attaching to the STDIN/STDOUT/STDERR streams of containers and container execs is [fundamentally flawed:](https://github.com/moby/moby/pull/45052#issuecomment-1492487126) the client closing its write side of the socket is used to signal EOF on STDIN, leaving no portable/reliable way for a client to signal its intention to detach from STDOUT/STDERR.

# Background

Currently, `/exec/{}/start` optionally accepts an `Upgrade` request header. The header value is ignored, and when present the Engine will respond with `HTTP/1.1 101 UPGRADED` and an `Upgrade: tcp` header. Otherwise it responds with `200 OK`. The connection is hijacked either way and is used as a full-duplex socket for passing STDIN/STDOUT/STDERR streams between the client and Engine.

# Proposal

I propose a new and improved protocol for container streams. The general idea is to separate the STDIN stream into a separate socket.

## `/exec/{}/start`

```
POST /exec/.../start HTTP/1.1
Connection: Upgrade
Upgrade: docker-container-attach/1
Content-type: application/json
Content-length: 13

{"Tty": true}

HTTP/1.1 101 Switching Protocols
Upgrade: docker-container-attach/1
Connection: Upgrade
Link: </exec/.../stdin>; rel="stdin"
```

- No Content-type header; the negotiated upgrade protocol is sufficient to indicate the wire format.
  - There is only one wire format for Engine -> Client data, which supports multiplexing of multiple concurrent data streams. This format is used unconditionally, even when `{"Tty": false}`, to allow for the transmission of out-of-band data such as the `stdcopy.Systemerr` stream.
  - For argument's sake, let's say that [`stdcopy`](https://github.com/moby/moby/blob/489543cd25db60f98a2892cb21c2071fd3be3907/pkg/stdcopy/stdcopy.go) is the format.
- The client can terminate the exec session by closing its side of the connection.
- The client MUST NOT write to the socket. Doing so is a protocol error; the Engine MAY terminate the session if a client does so.
  - A future version of the `docker-container-attach` protocol (negotiated with HTTP Upgrade) could utilize Client -> Engine writes for out-of-band communications of some sort
- The client must open the stdin stream separately. The presence of a `rel="stdin"` Link response header signals that an STDIN stream is available for attaching.

### The STDIN stream

```
POST /exec/.../stdin HTTP/1.1
Connection: Upgrade
Upgrade: docker-container-stdin/1

HTTP/1.1 101 Switching Protocols
Upgrade: docker-container-stdin/1
Connection: Upgrade
```

- After the upgrade handshake completes, any writes to the socket by the client translate to writes to the stdin stream.
- The client closing its side of the socket signals EOF. Subsequent attempts to open the STDIN stream fail.
- The Engine will close the connection to signal that a write failed with `EPIPE` or that the exec or attach session has ended.
- The Engine may write a message to the socket before closing the connection which describes the reason for closing the connection. The message is UTF-8 text.

## `/container/{}/attach`

Same basic protocol as exec above, with slight adjustments to account for the differences in semantics.

- Client closing the socket only signals that it wants to detach from the container.
  The container lifecycle, or the lifecycle of its streams, are unaffected.
- Client can attach and detach from the input stream and output streams independently.

### `/container/{}/attach-stdin`

- Client closing the socket only signals that it wants to detach from the container.
  The STDIN stream of the container process remains open.

**Comment by corhere** [comment]
@tianon cheekily suggested unifying execs with containers fully so all the same primitives work with both. I think it could really work! In that hypothetical world, an exec and a container would just be two flavours of process, much like `containerd.Process` and `containerd.Task`. To minimize differences we'd need to allow the stdout/stderr of execs to be logged, and the stdout/stderr of containers to bypass logging.

As far as the improved stream protocol proposal above goes, exec std-streams would become just like container std-streams: detachable and reattachable. A new API would be needed to close the write side of stdin.

**Reply by tianon** [reply]
It's essentially `--net container:xxx`, `--pid container:xxx`, etc but with a brand new `--mountns container:xxx` :grin:

It gets a little weird conceptually with the other namespaces that by default in `docker run` _aren't_ shared; would that mean you could technically "exec" into an existing container but have the new process be in a different network namespace?  PID namespace? :grimacing:

**Reply by corhere** [reply]
You're describing a Pod without a pause container!

I was thinking more along the lines of unifying the APIs used to manipulate _running_ containers and execs, but execs would still be processes spawned in all the same namespaces as the container pid1, same as today. Hence my analogy with `containerd.Process` and `containerd.Task`. Actually, that's pretty much literally what I am picturing: Engine API analogues for the whole [`containerd.Process`](https://pkg.go.dev/github.com/containerd/containerd#Process) interface which would operate on containers and execs interchangeably. Ignoring the `docker/cli` UX, I think it would make sense to continue having a distinct API for starting an exec in a container, directly analogous to `(containerd.Task).Exec()`.
